### ISSUE TITLE
You cannot call a method on a null-valued expression

### ISSUE THREAD

**Issue by collinbarrett** [zunderscore/chocolatey-packages#3]
Likely related to @zunderscore 's Disqus comments on the [choco repo page](http://disq.us/p/26daz6d).

A new version of Cascadia released recently. When I tried to update this package with choco, I got something like the error below. I then `choco uninstall`ed it and then tried to do a fresh `choco install`, and still get the error below.

Any ideas? Thanks for maintaining!

```
C:\WINDOWS\system32>choco install cascadiacodepl -y
Chocolatey v0.10.15
Installing the following packages:
cascadiacodepl
By installing you accept licenses for the packages.
Progress: Downloading cascadiacodepl 2005.15... 100%

cascadiacodepl v2005.15 [Approved]
cascadiacodepl package files install completed. Performing other installation steps.
Attempting to uninstall previous version of Cascadia Code PL font...
WARNING: An error occured removing 'CascadiaCodePL.ttf'.
ERROR: You cannot call a method on a null-valued expression.
The install of cascadiacodepl was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.

Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - cascadiacodepl (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
```

Log:
```
2020-05-19 08:05:16,441 9500 [DEBUG] - ---------------------------Script Execution---------------------------
2020-05-19 08:05:16,443 9500 [DEBUG] - Running 'ChocolateyScriptRunner' for cascadiacodepl v2005.15 with packageScript 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1', packageFolder:'C:\ProgramData\chocolatey\lib\cascadiacodepl', installArguments: '', packageParameters: '',
2020-05-19 08:05:16,448 9500 [DEBUG] - Running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'
2020-05-19 08:05:16,451 9500 [INFO ] - Attempting to uninstall previous version of Cascadia Code PL font...
2020-05-19 08:05:16,513 9500 [DEBUG] - 
   using System;
   using System.Collections.Generic;
   using System.Text;
   using System.IO;
   using System.Runtime.InteropServices;

   namespace FontResource {
       public class AddRemoveFonts {
           private static IntPtr HWND_BROADCAST = new IntPtr(0xffff);

           [DllImport("gdi32.dll")]
           static extern int AddFontResource(string lpFilename);

           [DllImport("gdi32.dll")]
           static extern int RemoveFontResource(string lpFileName);

           [return: MarshalAs(UnmanagedType.Bool)]
           [DllImport("user32.dll", SetLastError = true)]
           private static extern bool PostMessage(IntPtr hWnd, WM Msg, IntPtr wParam, IntPtr lParam);

           public enum WM : uint {
               FONTCHANGE = 0x001D
           }

           public static int AddFont(string fontFilePath) {
               FileInfo fontFile = new FileInfo(fontFilePath);
               if (!fontFile.Exists) { return 0; }
               try {
                   int retVal = AddFontResource(fontFilePath);
                   bool posted = PostMessage(HWND_BROADCAST, WM.FONTCHANGE, IntPtr.Zero, IntPtr.Zero);
                   return retVal;
               }
               catch { return 0; }
           }

           public static int RemoveFont(string fontFileName) {
               try {
                   int retVal = RemoveFontResource(fontFileName);
                   bool posted = PostMessage(HWND_BROADCAST, WM.FONTCHANGE, IntPtr.Zero, IntPtr.Zero);
                   return retVal;
               }
               catch { return 0; }
           }
       }
   }
2020-05-19 08:05:16,768 9500 [WARN ] - WARNING: An error occured removing 'CascadiaCodePL.ttf'.
2020-05-19 08:05:16,782 9500 [ERROR] - ERROR: You cannot call a method on a null-valued expression.
2020-05-19 08:05:16,785 9500 [DEBUG] - Built-in PowerShell host called with ['[System.Threading.Thread]::CurrentThread.CurrentCulture = '';[System.Threading.Thread]::CurrentThread.CurrentUICulture = ''; & import-module -name 'C:\ProgramData\chocolatey\helpers\chocolateyInstaller.psm1'; & 'C:\ProgramData\chocolatey\helpers\chocolateyScriptRunner.ps1' -packageScript 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1' -installArguments '' -packageParameters '''] exited with '-1'.
2020-05-19 08:05:16,787 9500 [DEBUG] - Calling command ['"C:\WINDOWS\System32\shutdown.exe" /a']
2020-05-19 08:05:16,884 9500 [DEBUG] - Command ['"C:\WINDOWS\System32\shutdown.exe" /a'] exited with '1116'
2020-05-19 08:05:16,973 9500 [DEBUG] - Capturing package files in 'C:\ProgramData\chocolatey\lib\cascadiacodepl'
2020-05-19 08:05:16,975 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\cascadiacodepl.nupkg'
  with checksum 'EC360018747EB197A32D15A1E9E282BE'
2020-05-19 08:05:16,975 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\cascadiacodepl.nuspec'
  with checksum '87799D2F873803D5D8C0438F61550519'
2020-05-19 08:05:16,977 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\files\CascadiaCodePL.ttf'
  with checksum 'FE5B88167E5951CC780EC6354E7564F3'
2020-05-19 08:05:16,978 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'
  with checksum '26858AD96C14802FD072D5F0CDA592D9'
2020-05-19 08:05:16,978 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyuninstall.ps1'
  with checksum '0E2A561C754BADB3108B317BA9A41C24'
2020-05-19 08:05:16,979 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\LICENSE.txt'
  with checksum '1DA97C42681B601447775D1051FBCDF9'
2020-05-19 08:05:16,980 9500 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\VERIFICATION.txt'
  with checksum '11A5318EF948414742C4AC08FD016F15'
2020-05-19 08:05:16,990 9500 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\cascadiacodepl.2005.15\.arguments".
2020-05-19 08:05:16,991 9500 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\cascadiacodepl.2005.15\.extra".
2020-05-19 08:05:16,992 9500 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\cascadiacodepl.2005.15\.version".
2020-05-19 08:05:16,992 9500 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\cascadiacodepl.2005.15\.sxs".
2020-05-19 08:05:16,993 9500 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\cascadiacodepl.2005.15\.pin".
2020-05-19 08:05:16,994 9500 [DEBUG] - Sending message 'HandlePackageResultCompletedMessage' out if there are subscribers...
2020-05-19 08:05:16,995 9500 [ERROR] - The install of cascadiacodepl was NOT successful.
2020-05-19 08:05:16,998 9500 [ERROR] - Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
2020-05-19 08:05:17,000 9500 [DEBUG] - Moving 'C:\ProgramData\chocolatey\lib\cascadiacodepl'
 to 'C:\ProgramData\chocolatey\lib-bad\cascadiacodepl'
2020-05-19 08:05:19,004 9500 [DEBUG] - Attempting to delete file "C:\Users\cbarre7\AppData\Local\NuGet\Cache\cascadiacodepl.2005.15.nupkg".
2020-05-19 08:05:19,011 9500 [WARN ] - 
Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
2020-05-19 08:05:19,013 9500 [INFO ] - 
2020-05-19 08:05:19,014 9500 [ERROR] - Failures
2020-05-19 08:05:19,015 9500 [ERROR] -  - cascadiacodepl (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
2020-05-19 08:05:19,017 9500 [DEBUG] - Sending message 'PostRunMessage' out if there are subscribers...
2020-05-19 08:05:19,033 9500 [DEBUG] - Exiting with -1
```

**Comment by zunderscore** [comment]
Hey @collinbarrett, thanks for using the packages!

This looks like it may be an issue with code being called from the [chocolatey-font-helpers.extension](https://github.com/bcurran3/ChocolateyPackages/tree/master/chocolatey-font-helpers.extension) package that the Cascadia font packages use to install/uninstall fonts from the system. Best I can figure is to try rebooting and running the install again, making sure you're running PowerShell as administrator when you do.

@bcurran3, I imagine you've seen this before. Any further advice?

**Comment by zunderscore** [comment]
Getting more reports of this since the 2007.01 release. @teknowledgist, since you were heavily involved with version 0.0.2 of the font helpers extension package, any insight you might have here?

**Comment by bcurran3** [comment]
I'm not sure what is going on, but it looks like a problem deleting/removing the originally installed font. I'd manually delete the font and then reinstall via Chocolatey. I'm guessing that will take care of the immediate problem. 

Teknowledgist, I'm sure can fix the script. I can look into it in the future when time allows. I have not scene the problem before. (I have a few fonts installed and really only for past testing.)

**Comment by collinbarrett** [comment]
I just saw it when trying to update to 2007.01 again as well. More info:

- I had `v2005.15` installed via `choco` on the machine.
- I ran `choco upgrade all -y` in admin terminal and observed:

```
You have cascadiacodepl v2005.15 installed. Version 2007.01 is available based on your source(s).
Progress: Downloading cascadiacodepl 2007.01... 100%

cascadiacodepl v2007.01 [Approved]
cascadiacodepl package files upgrade completed. Performing other installation steps.
Attempting to uninstall previous version of Cascadia Code PL font...
Cannot remove item C:\Windows\Fonts\CascadiaCodePL.ttf: Access to the path 'C:\Windows\Fonts\CascadiaCodePL.ttf' is denied.
Previous version of Cascadia Code PL font uninstalled successfully
Installing Cascadia Code PL font...
WARNING: An error occured installing 'C:\ProgramData\chocolatey\lib\cascadiacodepl\files\CascadiaCodePL.ttf'
ERROR: You cannot call a method on a null-valued expression.
The upgrade of cascadiacodepl was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
```

- I checked my Windows fonts, and I do not have Cascadia Code PL installed anymore (I do have some other Cascadias installed as a part of the new Windows Terminal).
- I then tried `choco uninstall cascadiacodepl` and observed:

```
Chocolatey v0.10.15
Uninstalling the following packages:
cascadiacodepl

cascadiacodepl v2005.15
The package cascadiacodepl wants to run 'chocolateyuninstall.ps1'.
Note: If you don't run this script, the installation will fail.
Note: To confirm automatically next time, use '-y' or consider:
choco feature enable -n allowGlobalConfirmation
Do you want to run the script?([Y]es/[A]ll - yes to all/[N]o/[P]rint): y

Uninstalling legacy Cascadia Code PL font...
WARNING: An error occured removing 'CascadiaPL.ttf'.
ERROR: You cannot call a method on a null-valued expression.
cascadiacodepl uninstall not successful.
Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyuninstall.ps1'.
 See log for details.
cascadiacodepl not uninstalled. An error occurred during uninstall:
 cascadiacodepl uninstall not successful.

Chocolatey uninstalled 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - cascadiacodepl (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyuninstall.ps1'.
 See log for details.

If a package uninstall is failing and/or you've already uninstalled the
 software outside of Chocolatey, you can attempt to run the command
 with `-n` to skip running a chocolateyUninstall script, additionally
 adding `--skip-autouninstaller` to skip an attempt to automatically
 remove system-installed software. This will only remove the packaging
 files and not things like software installed to Programs and Features.

If a package is failing because it is a dependency of another package
 or packages, then you may first need to consider if it needs removed
 as it is typically installed as a dependency for a reason. If you
 decide that you still want to remove it, head into
 `$env:ChocolateyInstall\lib` and find the package folder you want
 removed. Then delete the folder for the package. This option should
 only be used as a last resort.
```

- I then ran `choco uninstall cascadiacodepl -n` which seemed to have succeeded.
- I then ran `choco install cascadiacodepl -y` and observed again:

```
Chocolatey v0.10.15
Installing the following packages:
cascadiacodepl
By installing you accept licenses for the packages.
Progress: Downloading cascadiacodepl 2007.01... 100%

cascadiacodepl v2007.01 [Approved]
cascadiacodepl package files install completed. Performing other installation steps.
Attempting to uninstall previous version of Cascadia Code PL font...
WARNING: An error occured removing 'CascadiaCodePL.ttf'.
ERROR: You cannot call a method on a null-valued expression.
The install of cascadiacodepl was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.

Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - cascadiacodepl (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
```

Is there some other way I should try to manually force remove the old version of the font before trying to `choco install` again?

**Comment by teknowledgist** [comment]
Hi, @collinbarrett  I apologize for not taking this up sooner.  I couldn't at first, and then lost track of it.

I see a couple potential issues, and I'm not sure what is causing the first one you posted about...

When I try to test a basic install/uninstall of the current package, the uninstall is failing because the "ChocolateyUninstall.ps1" script is trying to remove the "CascadiaPL.ttf" font (and not "CascadiaCodePL.ttf") in this rather confusing section:

````
if ("CascadiaPL.ttf" -ne "CascadiaCodePL.ttf") {
    Write-Output "Uninstalling legacy Cascadia Code PL font..."
    
    if ((Uninstall-ChocolateyFont "CascadiaPL.ttf") -eq 1) {
        Write-Error "Error uninstalling legacy Cascadia Code PL font"
    }
````

The removal fails because the file doesn't exist, and then the `Write-Error` produces a terminating error due to the `$ErrorActionPreference = "Stop"`.

I'm not sure why you are getting the `Access to the path 'C:\Windows\Fonts\CascadiaCodePL.ttf' is denied.` message, but I suspect that is a different problem.

I'm also unclear why the `ERROR: You cannot call a method on a null-valued expression.` is occurring.  If the font file didn't exist, a message to that effect would come out as in the above case.  In order to get the `WARNING: An error occurred removing 'CascadiaCodePL.ttf'.` message, something has to fail in here:

````
         $retVal = [FontResource.AddRemoveFonts]::RemoveFont($FontPath)
         if ($retVal -eq 0) {
            Throw "Font resource, '$FontPath', removal failed"
         } else {
            Write-Verbose "Font resource, '$($File.FullName)', removed successfully"
            Remove-ItemProperty -Path $fontRegistryPath -Name "$FontName" -Force
            Remove-Item $FontPath -Force
            $successes++
         }
````

but the only thing in there that can cause `ERROR: You cannot call a method on a null-valued expression.` (that I know of) is the `$File.FullName` which should work since the file exists.  It seems contradictory to my eye, but I am not yet giving up trying to figure out what is going on though.  

IMO, it would be smarter for your package to put the pre-install removal stuff into a [ChocolateyBeforeModify.ps1](https://chocolatey.org/docs/create-packages#upgrading) script to check if the font file exists before uninstalling even though the code should work if it doesn't.  A true *install* doesn't have anything to remove, and an *update* would remove it before the "cleaner" ChocolateyInstall.ps1 runs.

**Comment by bcurran3** [comment]
Collin,

The big problem with ChocolateyBeforeModify.ps1 (and why I never use it) is that it won't run until the NEXT package is released; i.e. If a 0.0.3 of this package comes out (with a ChocolateyBeforeModify.ps1),  ChocolateyBeforeModify.ps1 won't run until 0.0.4 is released. Use of it is under the expectation that there will be further upgrade package versions.

This package was always considered a one and done... then @teknowledgist came along and made it better. :)

**Comment by teknowledgist** [comment]
@bcurran3 I don't entirely disagree.  That problem frustrates me in some of my packages too.  An alternative to the ChocolateyBeforeModify.ps1 script is to check for a previous install in the ChocolateyInstall.ps1 script rather than assume it's there and try to ignore the failed removal when it isn't there.

**Comment by MortenBoysen** [comment]
I saw the same problem. Three of the four cascadia packages upgraded correctly. The only one that failed was Cascadia Code PL, which was also the font used in the terminal running the install command. Can it be a problem with uninstalling a font that is in use?

**Comment by teknowledgist** [comment]
It's possible that an in-use font could be locked and cause a permission denied error.

**Comment by AlanMcBee** [comment]
Could this be the issue?

```
PSC C:\ProgramData\chocolatey\lib\cascadiacodepl\tools »  .\chocolateyinstall.ps1
Attempting to uninstall previous version of Cascadia Code PL font...
WARNING: An error occured removing 'CascadiaCodePL.ttf'.
InvalidOperation: C:\ProgramData\chocolatey\extensions\chocolatey-font-helpers\Uninstall-ChocolateyFont.ps1:91
Line |
  91 |           Write-Warning "$($error[0].ToString())"
     |                            ~~~~~~~~~~~~~~~~~~~~
     | You cannot call a method on a null-valued expression.

WARNING: Error uninstalling previous version of Cascadia Code PL font
Installing Cascadia Code PL font...
Cascadia Code PL font installed successfully
```

Update 2020-09-05
That line (91) comes from Uninstall-ChocolateyFont.ps1, btw. So maybe this is partially a chocolatey bug? I did a little more digging. In the error handlers for Install-ChocolateyFont.ps1 and Uninstall-ChocolateyFont.ps1, the exception is ignored, and only the $Error automatic variable is checked. I modified it to show the exception:

```
WPS C:\ProgramData\chocolatey\lib\cascadiacodepl\tools » .\chocolateyinstall.ps1
Attempting to uninstall previous version of Cascadia Code PL font...
WARNING: UNINSTALL: An error occured removing 'CascadiaCodePL.ttf'.
WARNING: 0
WARNING: Error uninstalling previous version of Cascadia Code PL font
Installing Cascadia Code PL font...
WARNING: INSTALL: An error occured installing 'C:\ProgramData\chocolatey\lib\cascadiacodepl\files\CascadiaCodePL.ttf'
WARNING: Exception: The requested operation cannot be performed on a file with a user-mapped section open.
WARNING: 0
C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1 : Error installing Cascadia Code PL font
At line:1 char:1
+ .\chocolateyinstall.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,chocolateyinstall.ps1
```

So that's that probably the root cause for this error.

As a workaround, I'm thinking. a logout/login cycle should fix it after the uninstall completes, or a shutdown/startup if there are multiple users.

(moments later) Confirmed. But, you do have to get the font uninstalled first.

**Comment by teknowledgist** [comment]
Oooh, interesting!  I should have noticed before that `WARNING: An error occurred removing 'CascadiaCodePL.ttf'."` is text explicitly output by *Uninstall-ChocolateyFont.ps1*, but the next bit, `ERROR: You cannot call a method on a null-valued expression.` is actually reporting an error in the error handling.  If it were also explicitly output, it would have started with `WARNING:`.

That's not a Chocolatey bug by any means.  I wouldn't call it a bug in *Uninstall-ChocolateyFont.ps1* either though, or at least not a bug in functionality.  Mostly, it's a limitation on error reporting.  

The error is that `$error[0]` can't be converted to a string because it is `$null`.  The only way it can be null is if the error occurs in  the `RemoveFontResource(fontFileName)` function call to `gdi32.dll` within the *New-FontResourceType.ps1* script. 

Almost certainly, the error coming from that function call could be captured and returned (which it is not -- hence the `$null`), but I don't know C# well enough to do that.  I stole the `AddRemoveFonts` class definition in that script from [somewhere else](https://blogs.technet.microsoft.com/deploymentguys/2010/12/04/adding-and-removing-fonts-with-windows-powershell/) and just cleaned it up a bit.  

I **think** what @AlanMcBee is showing is that the ultimate failure in this case is related to the font file being in use ("a file with a user-mapped section open").  Certainly, *chocolatey-font-helpers.extension* could be improved to provide a much more informative message for this kind of situation, but that will require C# skills beyond what I have at this point (or am likely to develop).

**Comment by zunderscore** [comment]
I might be able to look into it and maybe PR a patch into the font helpers extension to better handle the exception. At that point, the installation might still fail but at least users would have a better message as to why.

**Comment by zunderscore** [comment]
Alright, @teknowledgist, I've created a PR for your review :)

https://github.com/teknowledgist/Chocolatey-packages/pull/35

**Comment by tyler36** [comment]
I'm having the same error: ``` You cannot call a method on a null-valued expression.```

I managed to get the ```2007.01``` packaged uninstalled with ```choco uninstall cascadiacodepl -n```.

I can't seem to get ```cascadiacodepl v2008.25``` installed though:
```
❯ choco upgrade cascadiacodepl
Chocolatey v0.10.15
Upgrading the following packages:
cascadiacodepl
By upgrading you accept licenses for the packages.
cascadiacodepl is not installed. Installing...
Progress: Downloading cascadiacodepl 2008.25... 100%
Progress: Downloading cascadiacodepl 2008.25... 100%

cascadiacodepl v2008.25 [Approved]
cascadiacodepl package files upgrade completed. Performing other installation steps.
The package cascadiacodepl wants to run 'chocolateyinstall.ps1'.
Note: If you don't run this script, the installation will fail.
Note: To confirm automatically next time, use '-y' or consider:
choco feature enable -n allowGlobalConfirmation
Do you want to run the script?([Y]es/[A]ll - yes to all/[N]o/[P]rint): a

Attempting to uninstall previous version of Cascadia Code PL font...
WARNING: An error occured removing 'CascadiaCodePL.ttf'.
ERROR: You cannot call a method on a null-valued expression.
The upgrade of cascadiacodepl was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
```

OS: Win10

**Comment by teknowledgist** [comment]
@tyler36 Although you uninstalled the font package with `choco uninstall cascadiacodepl -n`, was the font file actually removed?  That `-n` at the end means the *ChocolateyUninstall.ps1* script wasn't run which means that Chocolatey removed the local package information even if the "application" (i.e. font) itself was still installed.  

If the *CascadiaCodePL.ttf* file didn't exist in the fonts folder, you should have gotten a `Warning: File Path 'C:\Windows\Fonts\CascadiaCodePL.ttf' does not exist!` message, and then the install proceeds.  The null-valued error should only be occurring if the file actually exists (and can't be removed using the DLL function called in the C# code).

**Comment by tyler36** [comment]
TLDR: Don't trust Font Settings and manually remove ```CascadiaCodePL.ttf```` via Powershell.

Thanks for the follow-up. I ran through the steps again. Indeed, there was some CascadiaCode fonts left behind. I uninstalled them via the font settings and restarted. But the same error occured.

I noticed there were other CascadiaCode fonts (normal and mono) which I tried to remove but it said they were a part of Windows Terminal. So I removed Windows Terminal (via choco). Removed other CascadiaCode fonts via Font Settings and restarted but same error.

Same error:
```
Attempting to uninstall previous version of Cascadia Code PL font...
WARNING: An error occured removing 'CascadiaCodePL.ttf'.
ERROR: You cannot call a method on a null-valued expression.
The install of cascadiacodepl was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\cascadiacodepl\tools\chocolateyinstall.ps1'.
 See log for details.
```

Re-read your comment and thought I'd double-check via Powershell. Yep, theres a "CascadiaCodePL.ttf" file but it wasn't showing up in Fonts Settings. Tried to remove it but it complained about permissions. Chenged to an Administrator Powershell session and remove it. Tried ```choco install cascadiacodepl``` and it worked! 🎉

Cheers again :)
